pkg/client/log/log.go

```diff
@@ -0,0 +1,617 @@
+// Copyright 2019-present Open Networking Foundation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/atomix/api/proto/atomix/headers"
+	api "github.com/atomix/api/proto/atomix/log"
+	"github.com/atomix/go-client/pkg/client/primitive"
+	"github.com/atomix/go-client/pkg/client/session"
+	"github.com/atomix/go-client/pkg/client/util"
+	"google.golang.org/grpc"
+)
+
+// Type is the log type
+const Type primitive.Type = "Log"
+
+// Index is the index of an entry
+type Index uint64
+
+// Version is the version of an entry
+type Version uint64
+
+// Client provides an API for creating IndexedMaps
+type Client interface {
+	// GetLog gets the log instance of the given name
+	GetLog(ctx context.Context, name string, opts ...session.Option) (Log, error)
+}
+
+// Log is a distributed log
+type Log interface {
+	primitive.Primitive
+
+	// Appends appends the given value to the end of the log
+	Append(ctx context.Context, value []byte) (*Entry, error)
+
+	// Get gets the value of the given index
+	Get(ctx context.Context, index int64, opts ...GetOption) (*Entry, error)
+
+	// FirstIndex gets the first index in the log
+	FirstIndex(ctx context.Context) (Index, error)
+
+	// LastIndex gets the last index in the log
+	LastIndex(ctx context.Context) (Index, error)
+
+	// PrevIndex gets the index before the given index
+	PrevIndex(ctx context.Context, index Index) (Index, error)
+
+	// NextIndex gets the index after the given index
+	NextIndex(ctx context.Context, index Index) (Index, error)
+
+	// FirstEntry gets the first entry in the log
+	FirstEntry(ctx context.Context) (*Entry, error)
+
+	// LastEntry gets the last entry in the log
+	LastEntry(ctx context.Context) (*Entry, error)
+
+	// PrevEntry gets the entry before the given index
+	PrevEntry(ctx context.Context, index Index) (*Entry, error)
+
+	// NextEntry gets the entry after the given index
+	NextEntry(ctx context.Context, index Index) (*Entry, error)
+
+	// Remove removes an entry from the log
+	Remove(ctx context.Context, index int64, opts ...RemoveOption) (*Entry, error)
+
+	// Len returns the number of entries in the log
+	Len(ctx context.Context) (int, error)
```

Comment:
It seems like `Size` would be more appropriate if this is the number of entries. `Len` feels like it would be equal to `LastIndex`.

Reply:
Done